### PR TITLE
[TypeScript] Add Bun

### PR DESCRIPTION
**Pull request by masfahru**

### CODE REVIEW

**Review by ImBIOS — APPROVED** [review]
LGTM! It's awesome to wait for the result.

close https://github.com/oven-sh/bun/issues/1077

Maybe @Jarred-Sumner and team (@oven-sh) want to review too.

**Comment by ImBIOS** [comment]
close https://github.com/oven-sh/bun/issues/1077

File: frameworks/TypeScript/bun/src/app.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+const serveFortune = async () => {
+  const fortunes = await db.fortunes();
+
+  fortunes.push({
+    id: 0,
+    message: 'Additional fortune added at request time.',
+  });
+
+  fortunes.sort((a, b) => a.message.localeCompare(b.message));
```
Comment (Jarred-Sumner) [inline frameworks/TypeScript/bun/src/app.ts:25]: don't use localeCompare

File: frameworks/TypeScript/bun/src/app.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  return new Response(
+    `<!DOCTYPE html><html><head><title>Fortunes</title></head><body><table><tr><th>id</th><th>message</th></tr>${html}</table></body></html>`,
+    headerHTML
+  );
+};
+
+const serveQuery = async (n: number) => {
+  const worldPromises = [];
+
+  for (let i = 0; i < n; i++) {
```
Comment (Jarred-Sumner) [inline frameworks/TypeScript/bun/src/app.ts:43]: push without initializing the length of the array?

File: frameworks/TypeScript/bun/src/app.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  const worlds = await Promise.all(worldPromises);
+  return new Response(JSON.stringify(worlds), headerJSON);
+};
+
+const serveUpdate = async (n: number) => {
+  const worldPromises = [];
+
+  for (let i = 0; i < n; i++) {
```
Comment (Jarred-Sumner) [inline frameworks/TypeScript/bun/src/app.ts:54]: push without initializing the length of the array?

File: frameworks/TypeScript/bun/src/app.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    worldPromises.push(db.find(generateRandomNumber()));
+  }
+
+  const worlds = await Promise.all(worldPromises);
+
+  await Promise.all(
+    worlds.map(async (world) => {
```
Comment (Jarred-Sumner) [inline frameworks/TypeScript/bun/src/app.ts:61]: unnecessary array allocation

File: frameworks/TypeScript/bun/src/app.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      world.randomNumber = generateRandomNumber();
+      await db.update(world);
+    })
+  );
+  return new Response(JSON.stringify(worlds), headerJSON);
+};
+
+const server = Bun.serve({
+  async fetch(req) {
+    const url = urlParser.parse(req.url);
```
Comment (Jarred-Sumner) [inline frameworks/TypeScript/bun/src/app.ts:71]: is the entire url parser necessary or just the first slash?

File: frameworks/TypeScript/bun/src/app.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        return serveJSON();
+        break;
+      case '/db':
+        return serveDB();
+        break;
+      case '/fortunes':
+        return serveFortune();
+        break;
+      case '/queries':
+        return serveQuery(parseQueries(url.search));
```
Comment (Jarred-Sumner) [inline frameworks/TypeScript/bun/src/app.ts:86]: why does this create another object?

File: frameworks/TypeScript/bun/src/database/postgres.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  password: 'benchmarkdbpass',
+  database: 'hello_world',
+  max: 1,
+});
+
+export const fortunes = async () => sql`SELECT id, message FROM fortune`;
+
+export const find = async (id: number) =>
+  sql`SELECT id, randomNumber FROM world WHERE id = ${id}`.then(
+    (arr) => arr[0]
```
Comment (Jarred-Sumner) [inline frameworks/TypeScript/bun/src/database/postgres.ts:15]: hoist the function to the top

File: frameworks/TypeScript/bun/src/headers.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  status: 200,
+  headers: {
+    'Content-Type': 'text/plain',
+    Server: 'bun',
+  },
+};
+
+export const headerJSON = {
+  status: 200,
+  headers: {
```
Comment (Jarred-Sumner) [inline frameworks/TypeScript/bun/src/headers.ts:11]: try creating these Header objects

you can skip the status code as it is 200 by default

File: frameworks/TypeScript/bun/src/utils.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
+export const generateRandomNumber = () => Math.floor(Math.random() * 9999) + 1;
+
+const regexGetN = /\?n=([^&]*)/;
+
+export const parseQueries = (n: string | null) => {
+  if (n) {
+    const match = regexGetN.exec(n);
```
Comment (Jarred-Sumner) [inline frameworks/TypeScript/bun/src/utils.ts:7]: don't use a regex for this?

File: frameworks/TypeScript/bun/src/app.ts [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  fortunes.push({
+    id: 0,
+    message: 'Additional fortune added at request time.',
+  });
+
+  fortunes.sort((a, b) => a.message.localeCompare(b.message));
+
+  let i = 0,
+    html = '';
+  for (; i < fortunes.length; i++)
```
Comment (Jarred-Sumner) [inline frameworks/TypeScript/bun/src/app.ts:29]: copy length into a local variable

File: frameworks/TypeScript/bun/src/app.ts [hunk 11]
```diff
@@ -0,0 +1,100 @@
+import * as db from './database/postgres.ts';
+import { headerPlainText, headerJSON, headerHTML } from './headers.ts';
+import { parseQueries, generateRandomNumber } from './utils.ts';
+
+const urlParser = require('fast-url-parser');
```
Comment (Jarred-Sumner) [inline frameworks/TypeScript/bun/src/app.ts:5]: did you actually test if this is faster than using `new URL`?

**Comment by Jarred-Sumner** [comment]
Bun should perform significantly better than what this benchmark indicates.

I'm not comfortable signing off on this in it's current state

**Comment by masfahru** [comment]
> Bun should perform significantly better than what this benchmark indicates.
> 
> I'm not comfortable signing off on this in it's current state

Alright, thank you for your confirmation @Jarred-Sumner , i'll close this PR

**Pull request closed without merging**